core/templates/filters/translate.pipe.ts

```diff
@@ -0,0 +1,149 @@
+  /**
+   * Variable to keep track of the last key sent for translation. If the
+   * previously sent key and params are same as the current one, the pipe
+   * doesn't request for translation and interpolation again.
+   */
+  lastKey: string;
+  /**
+   * Variable to keep track of the last param sent for translation. If the
+   * previously sent key and params are same as the current one, the pipe
+   * doesn't request for translation and interpolation again.
+   */
+  lastParams: Object;
+  onLangChange: Subscription;
+
+  constructor(
+    private translateService: TranslateService,
+    private ref: ChangeDetectorRef,
+    private utilsService: UtilsService
+  ) {}
+
+  /**
+   * Updates the value of interpolatedValue.
+   * @param {string} key - key for i18n translation
+   * @param {Object} interpolateParams - Params for interpolation
+   */
+  updateInterpolatedValue(key: string, interpolateParams?: Object): void {
+    this.interpolatedValue = this.translateService.getInterpolatedString(
+      key, interpolateParams);
+
+    // Storing the key to check if the key is same when the transform is invoked
+    // again.
+    this.lastKey = key;
+
+    // Storing the params to check if the params are same when the transform is
+    // invoked again.
+    this.lastParams = interpolateParams;
+    this.ref.markForCheck();
+  }
+
+  /**
+   * @param {string} key - key for i18n
+   * @param {Object} params - params for interpolation
+   * @returns {string} - Interpolated I18n value
+   */
+  transform(key: string, params?: Object | undefined): string {
+    if (!key || !key.length) {
+      return key;
+    }
+
+    // If the key and params are same, return the last stored value.
+    if (
+      key === this.lastKey &&
+      this.utilsService.isEquivalent(params, this.lastParams)) {
+      return this.interpolatedValue;
+    }
+
+    let interpolateParams: Object;
+    if (this.utilsService.isDefined(params) && !Array.isArray(params)) {
+      interpolateParams = params;
+    }
+
+    // Storing the key to check if the key is same when the transform is invoked
+    // again.
+    this.lastKey = key;
+
+    // Storing the params to check if the params are same when the transform is
+    // invoked again.
+    this.lastParams = params;
+
+    // Update the interpolated value.
+    this.updateInterpolatedValue(key, interpolateParams);
+
+    // If there is a subscription to onLangChange, unsubscribe.
+    this.disposeSubscriptions();
+    // Subscribe to onLangChange event, in case the language changes.
+    if (!this.onLangChange) {
+      this.onLangChange = this.translateService.onLangChange.subscribe(
+        (event: LangChangeEvent) => {
+          this.updateInterpolatedValue(key, interpolateParams);
+        }
+      );
+    }
+    return this.interpolatedValue;
+  }
+
+  /**
+   * Clean up any existing subscription to change events
+   */
+  private disposeSubscriptions(): void {
+    if (typeof this.onLangChange === 'undefined') {
```

Comment:
In general, here and elsewhere, prefer having a default value of null rather than undefined.

Also I don't think you need the typeof for the comparison check. Just use x === null.